lang/src/org/partiql/lang/eval/builtins/ExtractExprFunction.kt

```diff
@@ -38,22 +39,39 @@ internal class ExtractExprFunction(valueFactory: ExprValueFactory) : NullPropaga
 
     private fun Timestamp.minuteOffset() = (localOffset ?: 0) % SECONDS_PER_MINUTE
 
+    private fun extractedValue(datePart: DatePart, dateTimeValue: Any?) : Int {
+        return when (dateTimeValue) {
+            is Timestamp -> when (datePart) {
+                DatePart.YEAR -> dateTimeValue.year
+                DatePart.MONTH -> dateTimeValue.month
+                DatePart.DAY -> dateTimeValue.day
+                DatePart.HOUR -> dateTimeValue.hour
+                DatePart.MINUTE -> dateTimeValue.minute
+                DatePart.SECOND -> dateTimeValue.second
+                DatePart.TIMEZONE_HOUR -> dateTimeValue.hourOffset()
+                DatePart.TIMEZONE_MINUTE -> dateTimeValue.minuteOffset()
+            }
+            is LocalDate -> when (datePart) {
+                DatePart.YEAR -> dateTimeValue.year
+                DatePart.MONTH -> dateTimeValue.monthValue
+                DatePart.DAY -> dateTimeValue.dayOfMonth
+                DatePart.TIMEZONE_HOUR,
+                DatePart.TIMEZONE_MINUTE -> errNoContext("Timestamp unit ${datePart.name.toLowerCase()} not supported for DATE type", internal = false)
+                else -> 0
+            }
+            else         -> errNoContext("Expected date or timestamp: $dateTimeValue", internal = false)
+        }
+    }
```

Comment:
I think this should be two separate functions, one for Ion's `Timestamp` and one for `LocalDate`

Reply:
Indeed. Thanks. Done.